[redHat-arko] iP

arkobrat opened the pull request:
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Vincent frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s2.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is FREE!

Features:

 - [x] Managing tasks
 - [ ] Managing deadlines (coming soon)
 - [ ] Reminders (coming soon)

---

If you Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

arkobrat committed:
Added Greet, Echo, Exit

arkobrat committed:
Added Add, List

arkobrat committed:
Added Mark as Done/Undone

arkobrat committed:
Added ToDos, Events, Deadlines

arkobrat committed:
Added Automated Text UI Testing

arkobrat committed:
Handled incorrect input and empty todo errors

arkobrat committed:
Added Delete functionality

arkobrat committed:
Added Delete functionality

arkobrat committed:
Add Save functionality

arkobrat committed:
Add Date support for Deadlines

arkobrat committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java

arkobrat committed:
Add more OOP

arkobrat committed:
Organize into Packages

arkobrat committed:
Merge remote-tracking branch 'origin/add-gradle-support'

arkobrat committed:
Use Gradle

arkobrat committed:
Add JUnit Tests

arkobrat committed:
Add JavaDoc comments

arkobrat committed:
Follow the Coding Standard

arkobrat committed:
Add Find

arkobrat committed:
Merge remote-tracking branch 'origin/branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Storage.java

arkobrat committed:
Merge remote-tracking branch 'origin/branch-Level-9'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/Ui.java

arkobrat committed:
Resolve conflicts manually

arkobrat committed:
Resolve conflicts manually

inas0ng commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,49 @@
+
+    /**
+     * Parameterized constructor to create a Deadline
+     * @param description description of the task to be created
```
can consider having an empty line between description and parameter section for better readability in the javadoc comments :-)

inas0ng commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,96 @@
+     */
+    public void showDeleteMessage(Task deletedTask, int len) {
+        printFormatted("Noted. I've removed this task:\n" + deletedTask + "\nNow you have " +
+                len + " tasks in the list.");
```
should the line break happen before the plus sign? i.e.
printFormatted("Noted. I've removed this task:\n" + deletedTask + "\nNow you have "
                + len + " tasks in the list.");

inas0ng reviewed (commented):
your code looks good and clean! just pointed out some minor changes you may like to consider implementing :smile:

shirsho-12 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,49 @@
+
+public class Deadline extends Task {
+
+    protected LocalDate by;
```
I'd suggest using private instead of protected to follow the codestyle more closely.
```suggestion
    private LocalDate by;
```

shirsho-12 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,49 @@
+
+public class Event extends Task {
+    protected String from;
+    protected String to;
```
Similarly, you can replace protected with private here.

shirsho-12 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,157 @@
+     * @throws DukeException if an invalid command is entered
+     */
+    public void parse(String task) throws DukeException {
+        if (task.equals("list")) {
```
Would recommend using enums here.

shirsho-12 commented on src/test/java/duke/TodoTest.java:
```diff
@@ -0,0 +1,31 @@
+        assertEquals(t.toString(), "[T][ ] borrow book");
+    }
+    @Test
+    public void Test2() {
```
Nice use of testing.

shirsho-12 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    /**
+     * Prints a message on marking a Task as undone
```
Appreciate the extensive commenting.

shirsho-12 reviewed (commented):
Overall, I found your code well done and the extensive commenting was also great.

gwynethguo commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,157 @@
+     * Deletes the Task at a specified index
+     * @param index the index of the Task to be deleted
+     */
+    private void delete(int index) {
+        Task deletedTask = taskList.delete(index);
+        int len = taskList.size();
+        ui.showDeleteMessage(deletedTask, len);
+    }
```
Perhaps can change to deleteCommand instead? Because if the method name is only delete then people might assume that the method is to delete something related to the Parser itself :)

gwynethguo commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,157 @@
+     * @param task the String containing a single line of user input
+     * @throws DukeException if an invalid command is entered
+     */
+    public void parse(String task) throws DukeException {
+        if (task.equals("list")) {
+            showList();
+        } else if (task.startsWith("mark")) {
+            int index = Integer.parseInt(task.substring(5)) - 1;
+            mark(index);
+        } else if (task.startsWith("unmark")) {
+            int index = Integer.parseInt(task.substring(7)) - 1;
+            unmark(index);
+        } else if (task.startsWith("delete")) {
+            int index = Integer.parseInt(task.substring(7)) - 1;
+            delete(index);
+        } else if (task.startsWith("find")) {
+            find(task);
+        } else {
+            if (task.startsWith("todo")) {
+                addTodo(task);
+            } else if (task.startsWith("deadline")) {
+                addDeadline(task);
+            } else if (task.startsWith("event")) {
+                addEvent(task);
+            } else {
+                throw new DukeException("☹ OOPS!!! I'm sorry, but I don't know what that means :-(");
+            }
+            int len = taskList.size();
+            Task t = taskList.get(len - 1);
+            ui.showAddMessage(t, len);
+        }
+    }
```
Perhaps can consider using enums?

gwynethguo reviewed (commented):
Overall, your code follows the Java coding standard given in this module. Keep up the good work! :D

arkobrat committed:
Add GUI

arkobrat committed:
Fix bug in reading empty files

A NullPointerException is being thrown when the file where
tasks are stored ("duke.txt") is exists but is empty.

This wrongly terminates the execution of the program due
to some resources not being initialized.

Let's try to fix that by handling the DukeException that
gets thrown when there are no saved tasks to load, and
re-initializing resources to ensure that the application
runs.

arkobrat committed:
Fix bug in displaying an empty TaskList

An IndexOutOfBoundsException is thrown when the user tries
to list tasks from an empty TaskList.

This wrongly terminates the execution of the program.

Let's try to fix that by only displaying the tasks in the
TaskList when it is not empty.

arkobrat committed:
Use Assertions

arkobrat committed:
Improve Code Quality

arkobrat committed:
Merge pull request #2 from redHat-arko/branch-A-CodeQuality

Improve Code Quality

arkobrat committed:
Merge branch 'master' into branch-A-Assertions

arkobrat committed:
Merge pull request #3 from redHat-arko/branch-A-Assertions

Use Assertions

arkobrat committed:
Make task indexing more intuitive

After task deletion, the order in which the tasks are indexed changes,
which makes further operations on a specified index unintuitive.

Let's try to fix that by displaying the modified list after each addition/
deletion, thus making further operations much more intuitive.

arkobrat committed:
Use CheckStyle

arkobrat committed:
Make Find case-insensitive

arkobrat committed:
Add sorting functionality to Deadlines

arkobrat committed:
Use Enumerations

arkobrat committed:
Add Ui.png

arkobrat committed:
Merge branch 'branch-A-CodingStandard'

arkobrat committed:
Exit application on encountering exit command

arkobrat committed:
Add a User Guide

arkobrat committed:
Modify User Guide

arkobrat committed:
Update README.md

arkobrat committed:
Update README.md

arkobrat committed:
Use FMXL

arkobrat committed:
Add padding to text

arkobrat committed:
Set window title

arkobrat committed:
Fix checkstyle errors

arkobrat committed:
Update Ui.png
